File: sdk/python/examples/conformance.py
"""Local Mango HTTP-handler conformance; not evidence of a live model run."""

from __future__ import annotations

import asyncio
import os

from mango_sdk import APIError, AsyncMango, Mango


def main() -> None:
    url = os.environ["MANGO_SDK_TEST_URL"]
    key = os.environ["MANGO_SDK_TEST_KEY"]
    agents: list[str] = []
    environment_id: str | None = None
    session_id: str | None = None
    input_schema = {
        "type": "object", "properties": {"query": {"type": "string"}},
        "required": ["query"], "additionalProperties": False,
    }
    with Mango(base_url=url, api_key=key) as client:
        client.system.health()
        try:
            environment = client.environments.create(name="python-sdk-conformance", config={"type": "cloud"})
            environment_id = environment["id"]
            for suffix in ("one", "two"):
                agent = client.agents.create(name="python-sdk-" + suffix, model="sdk-conformance", tools=[{
                        "type": "custom", "name": "lookup", "description": "Look up a record",
                        "input_schema": input_schema,
                    }])
                agents.append(agent["id"])
                assert client.agents.retrieve(agent["id"])["name"] == "python-sdk-" + suffix
                assert agent["tools"][0]["input_schema"] == input_schema
            page = client.agents.list(limit=1)
            assert len(page["data"]) == 1 and page["next_page"]
            listed = {item["id"] for item in client.agents.iter(limit=1)}
            assert set(agents).issubset(listed)
            coordinator = client.agents.create(
                name="python-sdk-lead", model="sdk-conformance",
                multiagent={"type": "coordinator", "agents": [
                    {"type": "agent", "id": agents[0], "version": 1},
                    agents[1], {"type": "self"},
                    {"type": "advisor", "model": "review-model"},
                ]},
            )
            agents.append(coordinator["id"])
            assert coordinator["multiagent"]["agents"][0] == {
                "type": "agent", "id": agents[0], "version": 1,
            }
            assert coordinator["multiagent"]["agents"][-1] == {
                "type": "advisor", "model": "review-model",
            }
            session = client.sessions.create(agent={"type": "agent", "id": coordinator["id"]}, environment_id=environment_id)
            session_id = session["id"]
            with client.sessions.events.stream(session_id) as stream:
                sent = client.sessions.events.send(session_id, events=[{
                    "type": "user.message", "content": [{"type": "text", "text": "sdk test"}],
                }])
                assert len(sent["data"]) == 1
                observed = False
                for envelope in stream:
                    if envelope.data.get("id") == sent["data"][0]["id"]:
                        observed = True
                        break
                assert observed, "A ready subscription must receive the submitted event"
            history = list(client.sessions.events.iter(session_id, order="asc"))
            assert any(item["type"] == "user.message" for item in history)
            try:
                client.sessions.retrieve("sesn_python_missing")
            except APIError as error:
                assert error.status_code == 404
                assert error.type == "not_found_error"
                assert error.request_id
            else:
                raise AssertionError("Missing Session must return a typed 404 error")

            async def check_async() -> None:
                async with AsyncMango(base_url=url, api_key=key) as async_client:
                    await async_client.system.health()
                    assert session_id is not None
                    assert (await async_client.sessions.retrieve(session_id))["id"] == session_id

            asyncio.run(check_async())
        finally:
            if session_id:
                client.sessions.delete(session_id)
            for agent_id in agents:
                client.agents.archive(agent_id)
            if environment_id:
                client.environments.delete(environment_id)
    print("Python SDK local Mango HTTP conformance passed")


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/conformance.mjs
// For the repository's isolated real-HTTP conformance harness, not a live model demo.
import assert from 'node:assert/strict';
import { Mango, APIError } from '../dist/index.js';

const baseURL = process.env.MANGO_SDK_TEST_URL;
const apiKey = process.env.MANGO_SDK_TEST_KEY;
if (!baseURL || !apiKey) throw new Error('MANGO_SDK_TEST_URL and MANGO_SDK_TEST_KEY are required');
const client = new Mango({ baseURL, apiKey, timeoutMs: 10_000 });
const agents = [];
let environment;
let session;
try {
  await client.system.health();
  environment = await client.environments.create({ name: 'TypeScript conformance', config: { type: 'cloud' } });
  for (let index = 0; index < 2; index++) agents.push(await client.agents.create({ name: `TypeScript conformance ${index}`, model: 'sdk-conformance' }));
  assert.equal((await client.agents.retrieve(agents[0].id)).id, agents[0].id);
  const listed = [];
  for await (const item of client.agents.listItems({ limit: 1 })) listed.push(item.id);
  for (const agent of agents) assert.ok(listed.includes(agent.id));
  const coordinator = await client.agents.create({
    name: 'TypeScript lead', model: 'sdk-conformance',
    multiagent: { type: 'coordinator', agents: [
      { type: 'agent', id: agents[0].id, version: 1 }, agents[1].id,
      { type: 'self' }, { type: 'advisor', model: 'review-model' },
    ] },
  });
  agents.push(coordinator);
  assert.deepEqual(coordinator.multiagent.agents[0], { type: 'agent', id: agents[0].id, version: 1 });
  assert.deepEqual(coordinator.multiagent.agents.at(-1), { type: 'advisor', model: 'review-model' });
  session = await client.sessions.create({ agent: coordinator.id, environment_id: environment.id, title: 'TypeScript conformance' });
  assert.equal((await client.sessions.retrieve(session.id)).id, session.id);
  const live = await client.sessions.events.stream(session.id, {}, { signal: AbortSignal.timeout(10_000) });
  let batch;
  try {
    batch = await client.sessions.events.send(session.id, { events: [{ type: 'user.message', content: [{ type: 'text', text: 'SDK conformance' }] }] });
    assert.equal(batch.data.length, 1);
    let received = false;
    for await (const event of live) {
      if (event.id === batch.data[0].id) { received = true; break; }
    }
    assert.equal(received, true, 'ready live subscription must receive submitted input');
  } finally { await live.close(); }
  const events = [];
  for await (const event of client.sessions.events.listItems(session.id, { limit: 1 })) events.push(event);
  assert.ok(events.some(event => event.id === batch.data[0].id));
  await assert.rejects(client.sessions.retrieve('sesn_sdk_missing'), error => error instanceof APIError && error.status === 404 && error.type === 'not_found_error' && !!error.requestId);
} finally {
  // Cleanup failures must fail conformance rather than silently leaving resources.
  if (session) await client.sessions.delete(session.id);
  if (environment) await client.environments.delete(environment.id);
  for (const agent of agents) await client.agents.archive(agent.id);
}
console.log('TypeScript SDK real-HTTP conformance passed (test fakes, no model call)');
